---
title: "Перехват исключения при делении на 0"
---
<p><i><a href="https://github.com/MSDN-WhiteKnight/ruso-archive/">RuSO Archive</a></i></p>
<h1>Перехват исключения при делении на 0</h1>
<p><a href="https://ru.stackoverflow.com/questions/842745/%d0%9f%d0%b5%d1%80%d0%b5%d1%85%d0%b2%d0%b0%d1%82-%d0%b8%d1%81%d0%ba%d0%bb%d1%8e%d1%87%d0%b5%d0%bd%d0%b8%d1%8f-%d0%bf%d1%80%d0%b8-%d0%b4%d0%b5%d0%bb%d0%b5%d0%bd%d0%b8%d0%b8-%d0%bd%d0%b0-0">Source</a> - by <a href="https://ru.stackoverflow.com/users/255680/%d0%90%d0%bb%d0%b5%d0%ba%d1%81%d0%b0%d0%bd%d0%b4%d1%80">Александр</a></p>
<blockquote>
<p>Пытаюсь отловить исключение через catch/try: </p>

<pre><code>void TestCall()
{
    int b=0;
    try
    {

        int a = 1/b;        
    }
    catch(...)
    {
        test(1);
    }
}
</code></pre>

<p>Выражение <code>int a = 1/b;</code> заведомо вызывает исключение деления на 0 и я хоч его поймать в блоке catch. Но Visual C++ упорно вызывает ошибку на нем и на catch не перекидывает. В чем секрет?</p>

</blockquote>
<h2>Answer 842747</h2>
<p><a href="https://ru.stackoverflow.com/a/842747/">Source</a> - by <a href="https://ru.stackoverflow.com/users/272035/user7860670">user7860670</a></p>
<blockquote>
<p>Вы путаете исключения С++ и прочие исключения. В С++ исключения возникают исключительно при использовании оператора <code>throw</code>. Деление на 0 при использовании встроенного оператора <code>/</code> является неопределенным поведением:</p>

<blockquote>
  <p><strong>8.6 Multiplicative operators [expr.mul]</strong></p>
  
  <p>4 The binary / operator yields the quotient, and the binary % operator yields the remainder from the division of the first expression by the second. If the second operand of / or % is zero the behavior is undefined.</p>
</blockquote>

</blockquote>
<h2>Answer 843517</h2>
<p><a href="https://ru.stackoverflow.com/a/843517/">Source</a> - by <a href="https://ru.stackoverflow.com/users/240512/msdn-whiteknight">MSDN.WhiteKnight</a></p>
<blockquote>
<p>В Windows деление на ноль генерирует исключение SEH с кодом <em>STATUS_INTEGER_DIVIDE_BY_ZERO</em> (0xC0000094). Чтобы обрабатывать его в стиле исключений С++, можно установить в Visual C++ параметр компилятора <a href="https://docs.microsoft.com/en-us/cpp/build/reference/eh-exception-handling-model" rel="nofollow noreferrer">/EHa</a>, или установить фильтр необработанных исключений:</p>

<pre><code>#include &lt;stdlib.h&gt;
#include &lt;stdio.h&gt;
#include &lt;tchar.h&gt;
#include &lt;exception&gt;
#include &lt;windows.h&gt;

LONG WINAPI UnhandledExceptionFilter( struct _EXCEPTION_POINTERS *ExceptionInfo){

    if(ExceptionInfo-&gt;ExceptionRecord-&gt;ExceptionCode == STATUS_INTEGER_DIVIDE_BY_ZERO)
        throw std::exception("STATUS_INTEGER_DIVIDE_BY_ZERO");  

    return EXCEPTION_CONTINUE_SEARCH;
}

int _tmain(int argc, _TCHAR* argv[])
{   
    SetUnhandledExceptionFilter(&amp;UnhandledExceptionFilter); 

    int b=0;
    try
    {
        int a = 1/b;        
    }
    catch(std::exception ex)
    {       
        printf("Exception: %s\n",ex.what());
    }   

    return 0;
}
</code></pre>

<p>(Второй способ работает только при запуске без отладки, так как отладчик студии принудительно прерывает программу на SEH-исключениях.)</p>

</blockquote>
<hr/>
<p><i>Content is retrieved from StackExchange API. </i></p>
<p><i>Auto-generated by ruso-archive tools. </i></p>
